### PR TITLE
[JENKINS-33822] Create a new item using the keyboard

### PR DESCRIPTION
**Pull request by recena**
[JENKINS-33822](https://issues.jenkins-ci.org/browse/JENKINS-33822)

Ref.1 [Keyboard Accessibility](http://webaim.org/techniques/keyboard/tabindex)
Ref.2 [Keyboard-navigable JavaScript widgets](https://developer.mozilla.org/en-US/docs/Web/Accessibility/Keyboard-navigable_JavaScript_widgets)
Ref.3 [OpenAjax Alliance](http://www.oaa-accessibility.org/example/28)
Ref.4 [W3C HTML 5](https://www.w3.org/TR/html5/editing.html#editing)

This PR provides support for [WAI-ARIA](https://www.w3.org/WAI/intro/aria).

@reviewbybees especially @jtnord

I've recorded a [screencast](https://goo.gl/xNpn3D) in order to make easier the review of this PR.

### CODE REVIEW

File: war/src/main/js/add-item.js [hunk 1]
```diff
@@ -13,7 +13,7 @@ var getItems = function(){
 
 var jRoot = $('head').attr('data-rooturl');
 
-$.when(getItems()).done(function(data){
+$.when(getItems()).done(function(data) {
```
Comment (recena) [inline war/src/main/js/add-item.js]: Unrelated, only coding style.

File: war/src/main/js/add-item.js [hunk 2]
```diff
@@ -87,7 +87,7 @@ $.when(getItems()).done(function(data){
 
     function drawCategory(category) {
       var $category = $('<div/>').addClass('category').attr('id', 'j-add-item-type-' + cleanClassName(category.id));
-      var $items = $('<ul"/>').addClass('j-item-options');
+      var $items = $('<ul/>').addClass('j-item-options');
```
Comment (recena) [inline war/src/main/js/add-item.js]: Wrong markup introduced previously by me

File: war/src/main/js/add-item.js [hunk 3]
```diff
@@ -180,14 +194,14 @@ $.when(getItems()).done(function(data){
...
     // Init CopyFromField
-    $('input[name="from"]', '#createItem').focus(function() {
-      $('#createItem').find('input[type="radio"][value="copy"]').prop('checked', true);
-      $('.categories').find('.active').removeClass('active');
-    });
-
     $('input[name="from"]', '#createItem').blur(function() {
       if (getCopyFromValue() === '') {
-        $('#createItem').find('input[type="radio"][value="copy"]').prop('checked', false);
+        $('#createItem').find('input[type="radio"][value="copy"]').removeAttr('checked');
```
Comment (recena) [inline war/src/main/js/add-item.js]: Other way to do the same.

File: core/src/main/resources/hudson/model/View/newJob.jelly [hunk 4]
```diff
@@ -37,15 +37,15 @@ THE SOFTWARE.
...
+              <input name="name" id="name" type="text" tabindex="0" />
               <div class="input-help">&#187; ${%ItemName.help}</div>
               <div id="itemname-required" class="input-validation-message input-message-disabled">&#187; ${%ItemName.validation.required}</div>
               <div id="itemname-invalid" class="input-validation-message input-message-disabled"></div>
               <div id="itemtype-required" class="input-validation-message input-message-disabled">&#187; ${%ItemType.validation.required}</div>
             </div>
           </div>
 
-          <div class="categories flat" />
+          <div class="categories flat" role="radiogroup" aria-labelledby="Items"/>
```
Comment (amuniz) [inline core/src/main/resources/hudson/model/View/newJob.jelly]: [MDN Compatibility section](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-labelledby_attribute) seems to be TBD. I guess you tested in most popular browsers?

**Comment by amuniz** [comment]
It works fine in Chrome 48.0.2564.116, so :bee: for me.

File: core/src/main/resources/hudson/model/View/newJob.jelly (hunk 4 above)
Comment (recena) [inline core/src/main/resources/hudson/model/View/newJob.jelly]: *Add support information for common UA and AT product combinations* does not means it is not supported in its default web browser. TBD for others UA (*User agents*).

**Comment by recena** [comment]
@amuniz Thanks for your review. Did you try to create a new item and create a new item from other one?

**Comment by amuniz** [comment]
> Did you try to create a new item and create a new item from other one?

Do you mean "Copy"? Yes, I did.

**Comment by recena** [comment]
@amuniz Thanks!

**Comment by jtnord** [comment]
Seems to work locally on chrome.  I would expect [up/down keys](http://webaim.org/techniques/keyboard/) to navigate the list rather than tab - but at least it is navigable now without a mouse!

**Comment by recena** [comment]
@jtnord I'll include support for up/down keys in a separeted PR.

**Comment by recena** [comment]
@jtnord And IE?

**Comment by jtnord** [comment]
@recena IE - who uses that?  Yes it works there and edge.

Found a glicth though - if you enter a job name, but do not select any job type and click ok then the screen does not redraw to the error very well (the input box is hidden by the banner) - this is the same for non-keyboard navigation - so just another UX polish glitch

**Comment by recena** [comment]
@jtnord For that reason I hate the sticky elements!!!! Anyway, I'll addresse in the other PR join to up/down keys.

**Comment by jtnord** [comment]
:bee: based on testing - not code review.

**Comment by recena** [comment]
@reviewbybees done.

Thanks @amuniz and @jtnord for the testing.

**Comment by rsandell** [comment]
Navigating between the item types with `tab` seems a bit weird to me. `Tab` should only move you between distinct controls _(item type selector being one control and "copy from" being another)_ and then use the arrow keys to move between the individual types.

**Comment by recena** [comment]
@rsandell This is other option but modifies the default convention. More information in the links. Anyway, let me think about it.

**Comment by recena** [comment]
@daniel-beck This PR is ready to be merged.

**Comment by daniel-beck** [comment]
Tab-based navigation hiccups on hyperlinks in the item type descriptions, as for External Job. This did not happen on 1.x based radiobutton navigation (I could use arrow keys).

**Comment by recena** [comment]
@daniel-beck Yes, does not make sense when you are navigating through _items_ and it has defined the role **button** over the _items_

**Comment by recena** [comment]
@reviewbybees done

**Comment by recena** [comment]
I don't know way this PR is still waiting to be merged 😄

**Comment by recena** [comment]
@jenkinsci/code-reviewers your feedback is welcome!

**Comment by oleg-nenashev** [comment]
👍 according to the video.
I do not see obvious issues in the code

**Pull request merged**